core/dnsserver/server.go

```diff
@@ -328,12 +364,9 @@ func (s *Server) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 			}
 		}
 		off, end = dns.NextLabel(q, off)
-		if end {
-			break
-		}
 	}
 
-	if r.Question[0].Qtype == dns.TypeDS && dshandler != nil && dshandler.pluginChain != nil {
+	if dshandler != nil {
```

Comment:
This change is safe, because both `r.Question[0].Qtype == dns.TypeDS` and `dshandler.pluginChain != nil` need to pass before dshandler is set.